### DISCUSSION TITLE
HIP-796: Add Lockable Fractional Amounts of Fungible Tokens on Hedera

### DISCUSSION THREAD

**Discussion by Stephanie-Yi** [hiero-ledger/hiero-improvement-proposals#797] [Edits]
Opening up discussion for:
https://github.com/hashgraph/hedera-improvement-proposal/pull/796

Adds the ability to partition fungible and non-fungible tokens held by an account, and to lock a subset of tokens in an account or account partition, preventing those locked tokens from being transferred.

**Comment by david-bakin-sl** [comment]
This proposed HIP - for a very interesting and important feature - has a lot of moving parts and though the wording seems clear I can’t help but think it would be more solid if it included some diagrams and/or pseudocode to make things especially concrete.  There’s a lot of room for implementation fuzziness to sneak in to a big pile of words.  At least if there were diagrams and/or pseudocode a contradiction between the two could be noticed and addressed - it might be either the words or the diagram/code - whereas as it stands things might slip though the implementation.  And not just contradictions, but also holes.

In the same way, additional tables would be useful, e.g., which combination of parameters to APIs are _valid_ and which are not?  E.g., some of these APIs have 4 "buckets" on 2-dimensions:  fungible vs non-fungible, and token-definition-id vs partition-id.  It's not clear that _all_ parameters of some of these APIs are acceptable to all 4 buckets.  (And I'm _not_ talking about the difference between "amount" and "serial#", that's clear enough.)

**Comment by david-bakin-sl** [comment]
```
tokens: {[
    {
        token_id: 0.0.123456,
        balance: [
            {partition_id: 1, partition_id: 0.0.200001, balance: 10, locked: 5 },
            {partition_id: 3, partition_id: 0.0.200003, balance: 10, locked: 7 },
        ]
    }
]}
```
([here](https://github.com/TOKO-Network/hedera-improvement-proposal/blob/c77d4259160f547b1927d7f68b265792b46e71c9/HIP/hip-796.md?plain=1#L204))

The first `partition_id` should be `partition_number` as in the previous example.

**Comment by david-bakin-sl** [comment]
```
{
    id: 0.0.123456,
    ...
    partitions: [
        {partition_number:1, metadata: “ipfs://xxx”, partition_id: 0.0.200001},
        {partition_number:2, metadata: “ipfs://yyy”, partition_id: 0.0.200002},
        {partition_number:3, metadata: “ipfs://zzz”, partition_id: 0.0.200003}
    ]
}
```
([here](https://github.com/TOKO-Network/hedera-improvement-proposal/blob/c77d4259160f547b1927d7f68b265792b46e71c9/HIP/hip-796.md?plain=1#L187))

What is the purpose of `partition_number`?  `partition_id` is a sufficient primary key.  If `partition_number` is required how is it assigned (appears nowhere else in this document)?

**Reply by sam-at-luther** [reply]
`partition_number` is redundant, I believe only `partition_id` is now required, given that we record the `locked` amount on the partition_id.

**Comment by david-bakin-sl** [comment]
```
| Partition Key? | Partition-Move-Key? | Supply-Key? | Initial Supply? | Behavior                                                                                                                                                             |
| -------------- | ------------------- | ----------- | --------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| N              | Y                   | \*          | \*              | Acts like a normal token-def, no partition behavior                                                                                                                  |
| Y              | N                   | N           | Y               | No capabilities over a normal token-def, yet is more complex. Initial partition created with the initial supply                                                      |
| Y              | N                   | N           | N               | Not useful. Initial partition created with no supply and unable to mint                                                                                              |
```
([here](https://github.com/TOKO-Network/hedera-improvement-proposal/blob/c77d4259160f547b1927d7f68b265792b46e71c9/HIP/hip-796.md?plain=1#L345C1-L345C1))

I believe the first line of this table should read `N * * *`.

**Comment by Nana-EC** [comment]
Suggestions for missing content
1. **Lock-9**: As a lock-key smart contract holder, I want to  execute unlock/lock scenarios for NFT(s)/FT amounts in a users's account main balance or partition without requiring the user's signature.
2. Should `TokenLockTransactionBody` and `TokenUnLockTransactionBody` support an array of serial numbers to apply multiple NFTs at once? i.e. `repeated int64 serialNumbers = 4;` instead of `int64 serialNumber = 4;`?

**Comment by Nana-EC** [comment]
Updated Smart Contract section
```solidity

struct TokenPartitionInput {

    /**
     * The memo associated with the partition (UTF-8 encoding max 100 bytes)
     */
    string memo;

    /**
     * The publicly visible name of the partition. The partition name is specified as a Unicode
     * string. Its UTF-8 encoding cannot exceed 100 bytes, and cannot contain the 0 byte (NUL). 
     * This name is not unique within the ledger.
     */
    string name;

    /**
     * The token (either FUNGIBLE_COMMON or NON_FUNGIBLE_UNIQUE) that this partition is a part of.
     */
    address parentToken;
}

struct TokenPartitionInfo {
    /**
     * The creation time of the partition
     */
    int64 creationTime;

    /**
     * The memo associated with the partition (UTF-8 encoding max 100 bytes)
     */
    string memo;

    /**
     * The publicly visible name of the partition. The partition name is specified as a Unicode
     * string. Its UTF-8 encoding cannot exceed 100 bytes, and cannot contain the 0 byte (NUL). 
     * This name is not unique within the ledger.
     */
    string name;

    /**
     * The token (either FUNGIBLE_COMMON or NON_FUNGIBLE_UNIQUE) that this partition is a part of.
     */
    address parentToken;

    /**
     * The address of the token partition
     */
    address partition;
}

struct HederaToken {
	...

  // list of token partitions on this token
  TokenPartitionInfo[] partitions;
}

struct TokenKey {

    // bit field representing the key type. Keys of all types that have corresponding bits set to 1
    // will be created for the token.
    // ...
    // 7th bit: lockKey
    // 8th bit: partitionKey
    // 9th bit: partitionMoveKey
    // 10th bit: ignored
    uint keyType;

    // the value that will be set to the key type
    KeyValue key;
}

struct TokenCreateResponse {
	address partition;
}

struct TokenAmount {
	int64 amount;
	int64[] serialNumbers;
}

interface IHederaTokenService {
		...

		// new versions of existing functions
    function createFungibleToken(HederaToken memory token, int64 initialTotalSupply, int32 decimals) external payable returns (int64 responseCode, address tokenAddress, TokenCreateResponse memory tokenCreateResponse);
    function createFungibleTokenWithCustomFees(HederaToken memory token, int64 initialTotalSupply, int32 decimals, FixedFee[] memory fixedFees, FractionalFee[] memory fractionalFees) external payable returns (int64 responseCode, address tokenAddress, TokenCreateResponse memory tokenCreateResponse);
    function createNonFungibleToken(HederaToken memory token) external payable returns (int64 responseCode, address tokenAddress, TokenCreateResponse memory tokenCreateResponse);
    function createNonFungibleTokenWithCustomFees(HederaToken memory token, FixedFee[] memory fixedFees, RoyaltyFee[] memory royaltyFees) external payable returns (int64 responseCode, address tokenAddress, TokenCreateResponse memory tokenCreateResponse);
    
    // new functions
    function createTokenPartition(TokenPartitionInput memory tokenPartitionInput) external returns (uint256 responseCode, address tokenPartition);
    function deleteTokenPartition(address tokenPartition) external returns (uint256 responseCode);
    function getLockedTokenAmount(address token, address account) external returns (int64 responseCode, TokenAmount memory tokenAmount);
    function getTokenPartitionInfo(address token) external returns (int64 responseCode, TokenPartitionInfo memory tokenPartitionInfo);
    function isLocked(address token, address account, int64[] memory serialNumbers) external returns (int64 responseCode, bool locked);
    function lockToken(address token, address account, int64 amount, int64[] memory serialNumbers) external returns (uint256 responseCode);
    function unlockToken(address token, address account, int64 amount, int64[] memory serialNumbers) external returns (uint256 responseCode);
    function updateTokenPartition(address tokenPartition, TokenPartitionInfo memory tokenPartitionInfo) external returns (uint256 responseCode);
}

```

Along with the update hashes and signatures
```
Function hashes: Function signatures
0x53933c02: createFungibleToken((string,string,address,string,bool,int64,bool,(uint256,(bool,address,bytes,bytes,address))[],(int64,address,int64),(int64,string,string,address,address)[]),int64,int32)
0xb8178f33: createFungibleTokenWithCustomFees((string,string,address,string,bool,int64,bool,(uint256,(bool,address,bytes,bytes,address))[],(int64,address,int64),(int64,string,string,address,address)[]),int64,int32,(int64,address,bool,bool,address)[],(int64,int64,int64,int64,bool,address)[])
0xe6737645: createNonFungibleToken((string,string,address,string,bool,int64,bool,(uint256,(bool,address,bytes,bytes,address))[],(int64,address,int64),(int64,string,string,address,address)[]))
0x1f537c3f: createNonFungibleTokenWithCustomFees((string,string,address,string,bool,int64,bool,(uint256,(bool,address,bytes,bytes,address))[],(int64,address,int64),(int64,string,string,address,address)[]),(int64,address,bool,bool,address)[],(int64,int64,int64,address,bool,address)[])
0x0ec8b6d5: createTokenPartition((int64,string,string,address,address))
0x48ebe709: deleteTokenPartition(address)
0xadec7128: getLockedTokenAmount(address,address)
0x5f77bb56: getTokenPartitionInfo(address)
0x8612d049: isLocked(address,address)
0x40c1d8c1: lockToken(address,address,int64,int64[])
0x1ba86ed8: unlockToken(address,address,int64,int64[])
0xfaf56035: updateTokenPartition(address,(int64,string,string,address,address))
```

**Reply by sam-at-luther** [reply]
I belive this would also require a `moveToken` to move tokens across partitions, correct?

**Comment by HederaGraph** [comment]
PROPOSAL FOR PLANNING THE SEGMENTATION AND NUMBER OF CHARACTERS FOR WALLET CODES RELATED TO IDENTITY

Hedera's wallet token ID has a structure that is far superior to wallet codes on all other platforms
Displaying the wallet code in numbers is a huge advantage and benefit, because the display of the hederea wallet code will easily meet all identity requirements in all industries
- The structure of the hedera wallet code can easily be used for identification as a citizen identification number in countries, it is similar to a bank account, and can also be used as a mobile telecommunications system... and in most areas of life related to human identity

Based on these advantages, I would like to propose a proposal for the allocation of numbers according to the structure of the wallet code hedera <shardNum>.<realmNum>.<account> . as follows
1The initial planning with the <account> section will be limited by default to less than 15 numeric characters to apply to identity issues, or ID numbers, cbdc bank account numbers...., if the number exceeds 15 numeric characters, the segments in the <realmNum> section will continue to be developed
and with this approach, all tokens with a format of 15 numeric characters or more up to the end of 8 numeric bytes in the <account> section will be assigned a default token number for nfts, types of messenger topics....

As I observe at the present time, with typical examples such as IKAD learning certificate nfts, or for example Topic Messages
all exploit tokens or many types of nfts (junk, miscellaneous) are assigned to bank tokens with the user's Wallet token ID
Using resources like this is a waste, and in the future, there will be countless products mass production is assigned id tokens in such a wasteful way
On the other hand, if we know how to take advantage of and plan this problem, the power of hedera will be closer to real life, and will increase the value of hedera tokens many times over other block change wallet codes.
Imagine, hedera's ID token wallet will act as a cross-chain bridge, and users will transfer money to each other as conveniently as sending bank account numbers to each other in real life. Even tokens of the same type (for example ETH) will be very reluctant to transfer directly on the chain to each other, but instead, they will transfer through the hedera bridge with a very small fee in exchange for convenience and ease
Therefore, plan from the beginning, do not let wallet codes, identity accounts have too long characters.... At the same time, you should also reconsider how to assign nfts, Topic Messages... to long character wallet codes, in order to optimize IDtoken resources with numbers short characters for identity needs and user wallets

Please share and interact, send this suggestions and ideas to hedera developers
